nifi-nar-bundles/nifi-cdc/nifi-cdc-mysql-bundle/nifi-cdc-mysql-processors/src/main/java/org/apache/nifi/cdc/mysql/processors/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,1007 @@
+                lastStateUpdate = now;
+            }
+        } catch (IOException ioe) {
+            try {
+                // Perform some processor-level "rollback", then rollback the session
+                currentBinlogFile = xactBinlogFile == null ? "" : xactBinlogFile;
+                currentBinlogPosition = xactBinlogPosition;
+                currentSequenceId.set(xactSequenceId);
+                inTransaction = false;
+                stop(stateManager);
+                queue.clear();
+                currentSession.rollback();
+            } catch (Exception e) {
+                // Not much we can recover from here
+                log.warn("Error occurred during rollback", e);
+            }
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnStopped
+    public void onStopped(ProcessContext context) {
+        try {
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnShutdown
+    public void onShutdown(ProcessContext context) {
+        try {
+            // In case we get shutdown while still running, save off the current state, disconnect, and shut down gracefully
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    /**
+     * Get a list of hosts from a NiFi property, e.g.
+     *
+     * @param hostsString A comma-separated list of hosts (host:port,host2:port2, etc.)
+     * @return List of InetSocketAddresses for the hosts
+     */
+    private List<InetSocketAddress> getHosts(String hostsString) {
+
+        if (hostsString == null) {
+            return null;
+        }
+        final List<String> hostsSplit = Arrays.asList(hostsString.split(","));
+        List<InetSocketAddress> hostsList = new ArrayList<>();
+
+        for (String item : hostsSplit) {
+            String[] addresses = item.split(":");
+            if (addresses.length != 2) {
+                throw new ArrayIndexOutOfBoundsException("Not in host:port format");
+            }
+
+            hostsList.add(new InetSocketAddress(addresses[0].trim(), Integer.parseInt(addresses[1].trim())));
+        }
+        return hostsList;
+    }
+
+    protected void connect(List<InetSocketAddress> hosts, String username, String password, boolean createEnrichmentConnection,
+                           String driverLocation, String driverName, long connectTimeout) throws IOException {
+
+        int connectionAttempts = 0;
+        final int numHosts = hosts.size();
+        InetSocketAddress connectedHost = null;
+
+        while (connectedHost == null && connectionAttempts < numHosts) {
+            if (binlogClient == null) {
+
+                connectedHost = hosts.get(currentHost);
+                binlogClient = createBinlogClient(connectedHost.getHostString(), connectedHost.getPort(), username, password);
```

Comment:
I copied existing CaptureChangeMySQL processor and run two processors having the same configuration, then I got following error message:

```
2017-04-06 21:07:28,262 ERROR [Timer-Driven Process Thread-9] o.a.n.c.m.processors.CaptureChangeMySQL
com.github.shyiko.mysql.binlog.network.ServerException: A slave with the same server_uuid/server_id as this slave has connected to the master; the first event 'mysql-server-bin.000003' at 7778, the last event read from './mysql-server-bin.000003' at 123, the last byte read from './mysql-server-bin.000003' at 7778.
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:882) ~[mysql-binlog-connector-java-0.11.0.jar:0.11.0]
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) ~[mysql-binlog-connector-java-0.11.0.jar:0.11.0]
        at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) ~[mysql-binlog-connector-java-0.11.0.jar:0.11.0]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
```

BinLogClient has a method to set serverId. We might want to set unique one for each processor. Adding a processor property, not required, but let user set an unique serverId (Integer) within replication group would be nice.

Since this is not a critical issue, I'm fine do it in separate JIRA.